pkg/controller/statefulset/stateful_set_utils.go

```diff
@@ -335,16 +335,17 @@ func newRevision(set *appsv1alpha1.StatefulSet, revision int64, collisionCount *
 // ApplyRevision returns a new StatefulSet constructed by restoring the state in revision to set. If the returned error
 // is nil, the returned StatefulSet is valid.
 func ApplyRevision(set *appsv1alpha1.StatefulSet, revision *apps.ControllerRevision) (*appsv1alpha1.StatefulSet, error) {
-	clone := set.DeepCopy()
-	patched, err := strategicpatch.StrategicMergePatch([]byte(runtime.EncodeOrDie(patchCodec, clone)), revision.Data.Raw, clone)
+	patched, err := strategicpatch.StrategicMergePatch([]byte(runtime.EncodeOrDie(patchCodec, set)), revision.Data.Raw, set)
 	if err != nil {
 		return nil, err
 	}
-	err = json.Unmarshal(patched, clone)
+
+	restoredSet := &appsv1alpha1.StatefulSet{}
+	err = json.Unmarshal(patched, restoredSet)
```

Comment:
How about 
```
	var restoredSet *appsv1alpha1.StatefulSet
	err = json.Unmarshal(patched, &restoredSet)
```

Reply:
it will panic. `var restoredSet *appsv1alpha1.StatefulSet` equals `var restoredSet *appsv1alpha1.StatefulSet = nil`, and can not `json.Unmarshal(nil)`

Reply:
It will not. Just try it.

Reply:
@FillZpp 

it should be 
```go
    var restoredSet appsv1alpha1.StatefulSet
    err = json.Unmarshal(patched, &restoredSet)
```

or it will panic as @tcmichael mentioned.